### PR TITLE
Default configuration for vertx dns client

### PR DESCRIPTION
**Pull request by Sammers21**
Signed-off-by: Pavel Drankov <titantins@gmail.com>

A solution for https://github.com/eclipse/vert.x/issues/2239

### CODE REVIEW

**Comment by Sammers21** [comment]
@vietj , do we actually should have a test for that?

**Comment by vietj** [comment]
@Sammers21 of course! I recommend to configure the VertxOptions with the embedded test DNS server

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 1]
```diff
@@ -406,6 +409,15 @@ public DnsClient createDnsClient(int port, String host) {
     return new DnsClientImpl(this, port, host, DnsClientOptions.DEFAULT_QUERY_TIMEOUT);
   }
 
+  @Override
+  public DnsClient createDnsClient() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/VertxImpl.java]: actually it should use `DnsResolverProvider` so it can be configured using `VertxOptions#addressResolverOptions` (that's how we can test it).

Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/VertxImpl.java]: In `DnsResolverProvider`s constructor, we have a `serverList` so the list of DNS servers to use, but when creating DnsClient we are specifying only one sever. Should we create `DnsResolverProvider` with `addressResolverOptions` and then use the first element of serverList as a default DNS server with port=53?

Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/VertxImpl.java]: I have just updated the PR. Is that what you mean?

File: src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java [hunk 2]
```diff
@@ -62,10 +62,10 @@
   private final Vertx vertx;
   private final List<ResolverRegistration> resolvers = Collections.synchronizedList(new ArrayList<>());
   private AddressResolverGroup<InetSocketAddress> resolverGroup;
+  public final List<InetSocketAddress> serverList = new ArrayList<>();
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: what's the purpose of this ?

File: src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java [hunk 3]
```diff
@@ -85,7 +85,7 @@ public DnsResolverProvider(VertxImpl vertx, AddressResolverOptions options) {
         }
       }
     } else {
-      DnsServerAddressStream stream = DnsServerAddresses.defaultAddresses().stream();
+      DnsServerAddressStream stream = DefaultDnsServerAddressStreamProvider.defaultAddresses().stream();
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: why this change ?

File: src/main/java/io/vertx/core/impl/VertxImpl.java (hunk 1 above)
Comment (vietj) [reply src/main/java/io/vertx/core/impl/VertxImpl.java]: yes, we should add a method on DnsResolverProvider that provides the server address (the first one) and have only the stream nameServerAddresses to be a field of DnsResolverProvider with a method that returns it, so when building the client, it can use the first one rather than using a List

File: src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: since `DnsServerAddresses#defaultAddresses()` is deprecated

Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: it is recommended to use  `DefaultDnsServerAddressStreamProvider#defaultAddresses()`

File: src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: that is where we are storing DNS server addresses

Comment (Sammers21) [reply src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: so the first one will be used as a default one

**Comment by Sammers21** [comment]
@vietj , I have just added a test. So, what I m trying to achieve is forcing to use a fake DNS server as a default one and then check that the first element of the list returned by the _nameServerAddresses_ method is the fake server.

If something is wrong just let me know.

File: src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java (hunk 3 above)
Comment (vietj) [reply src/main/java/io/vertx/core/impl/resolver/DnsResolverProvider.java]: ok

File: src/main/java/io/vertx/core/Vertx.java [hunk 4]
```diff
@@ -220,6 +220,13 @@ static void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>> res
    */
   DnsClient createDnsClient(int port, String host);
 
+  /**
+   * Create a DNS client to connect to automatically chosen DNS server
```
Comment (vietj) [inline src/main/java/io/vertx/core/Vertx.java]: Create a DNS client to connect to the DNS server configured by {@link VertxOptions#getAddressResolverOptions()}

Comment (Sammers21) [reply src/main/java/io/vertx/core/Vertx.java]: changed

File: src/test/java/io/vertx/test/core/DNSTest.java [hunk 5]
```diff
@@ -63,6 +66,27 @@ public void testIllegalArguments() throws Exception {
     dnsServer.stop();
   }
 
+  @Test
+  public void testDefaultDnsServer() throws Exception {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/DNSTest.java]: why isn't the test using createDnsClient() ?

Comment (Sammers21) [reply src/test/java/io/vertx/test/core/DNSTest.java]: In the test, I have tried to test that the first DNS server in a list is our fake DNS server. 

I m agree with you. We should test it with direct _createDnsClient()_ call.

I have just made some changes. Is that correct?

**Comment by Sammers21** [comment]
@vietj , is it fine now?

**Comment by vietj** [comment]
@Sammers21 one last thing : we need to say in the javadoc that the DNS client takes the first configured resolver address

**Comment by Sammers21** [comment]
@vietj, done
